frontend/src/metabase/common/components/FilterPicker/CoordinateFilterPicker/CoordinateFilterPicker.tsx

```diff
@@ -1,55 +1,62 @@
 import { t } from "ttag";
 import { useState, useMemo } from "react";
-import { Box, Button, Flex, Text, NumberInput, Stack } from "metabase/ui";
-import * as Lib from "metabase-lib";
 
+import { Box, Button, Flex, NumberInput, Text, Stack } from "metabase/ui";
+import * as Lib from "metabase-lib";
 import FieldValuesWidget from "metabase/components/FieldValuesWidget";
 import Field from "metabase-lib/metadata/Field";
+
 import type { FilterPickerWidgetProps } from "../types";
+import { getAvailableOperatorOptions } from "../utils";
 import { BackButton } from "../BackButton";
 import { Header } from "../Header";
 import { Footer } from "../Footer";
-
-import { FilterOperatorPicker } from "../FilterOperatorPicker";
 import { FlexWithScroll } from "../FilterPicker.styled";
+import { FilterOperatorPicker } from "../FilterOperatorPicker";
 
+import { OPTIONS } from "./constants";
+import { findSecondColumn } from "./utils";
 import { CoordinateColumnSelect } from "./CoordinateColumnSelect";
 
-import { findSecondColumn, isCoordinateFilterValid } from "./utils";
-
-import { coordinateFilterValueCountMap } from "./constants";
-import type { CoordinateFilterValueCount } from "./types";
-
 export function CoordinateFilterPicker({
   query,
   stageIndex,
   column,
   filter,
-  onBack,
   onChange,
+  onBack,
 }: FilterPickerWidgetProps) {
   const columnName = Lib.displayInfo(query, stageIndex, column).longDisplayName;
   const filterParts = filter
     ? Lib.coordinateFilterParts(query, stageIndex, filter)
     : null;
 
-  const [operatorName, setOperatorName] =
-    useState<Lib.CoordinateFilterOperatorName>(
-      filterParts
-        ? filterParts.operator
-        : (Lib.defaultFilterOperatorName(
-            query,
-            stageIndex,
-            column,
-          ) as Lib.CoordinateFilterOperatorName),
-    );
-
-  const [values, setValues] = useState<number[]>(filterParts?.values ?? []);
+  const availableOperators = useMemo(
+    () => getAvailableOperatorOptions(query, stageIndex, column, OPTIONS),
+    [query, stageIndex, column],
+  );
 
-  const [column2, setColumn2] = useState<Lib.ColumnMetadata | null>(
+  const [operatorName, setOperatorName] = useState(
+    filterParts?.operator ?? "=",
+  );
+  const [values, setValues] = useState(filterParts?.values ?? []);
+  const [column2, setColumn2] = useState(
     findSecondColumn({ query, stageIndex, column, filter, operatorName }),
   );
 
+  const valueCount = useMemo(() => {
+    const option = availableOperators.find(
+      option => option.operator === operatorName,
+    );
+    return option?.valueCount ?? 0;
+  }, [availableOperators, operatorName]);
+
+  const isFilterValid = useMemo(() => {
+    return Number.isFinite(valueCount)
+      ? values.length === valueCount
+      : values.length >= 1;
+  }, [values, valueCount]);
```

Comment:
Two nits:
- Not sure this needs to be memoized, the useMemo function overhead might outweight this simple check
- for multi-value inputs (like between), it marks them as valid even if they only have 1 value, I think it's cuz the value is `[17, undefined]` which has length 2. For numbers we should filter to numeric values. (this bug was present previously too)

Reply:
> Not sure this needs to be memoized, the useMemo function overhead might outweight this simple check

Yeah, good point 👍 

> for multi-value inputs (like between), it marks them as valid even if they only have 1 value, I think it's cuz the value is [17, undefined] which has length 2. For numbers we should filter to numeric values. (this bug was present previously too)

OH, nice catch! Will fix this, thank you